### PR TITLE
[BEAM-8335] Add timestamp and duration to/from protos to Python SDK

### PR DESCRIPTION
**Pull request by rohdesamuel**
Adds extra utility in the Python SDK Timestamp object to convert between it and the well-known google.protobuf.timestamp.Timestamp object.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by rohdesamuel** [comment]
R: @lukecwik Hey Luke, are you able to review this please?

File: sdks/python/apache_beam/utils/timestamp.py [hunk 1]
```diff
@@ -140,6 +141,23 @@ def to_rfc3339(self):
     # Append 'Z' for UTC timezone.
     return self.to_utc_datetime().isoformat() + 'Z'
 
+  def to_proto(self):
+    """Returns the `google.protobuf.timestamp_pb2` representation."""
+    secs = self.micros // 1000000
+    nanos = (self.micros - (secs * 1000000)) * 1000
```
Comment (lukecwik) [inline sdks/python/apache_beam/utils/timestamp.py]:
```suggestion
    nanos = (self.micros % 1000000) * 1000
```

File: sdks/python/apache_beam/utils/timestamp.py [hunk 2]
```diff
@@ -140,6 +141,23 @@ def to_rfc3339(self):
...
+  @staticmethod
+  def from_proto(timestamp_proto):
+    """Creates a Timestamp from a `google.protobuf.timestamp_pb2`.
+
+    Note that the google has a sub-second resolution of nanoseconds whereas this
+    class has a resolution of microsends. This class will truncate the
+    nanosecond resolution down to the microsecond.
+    """
+    return Timestamp(seconds=timestamp_proto.seconds,
+                     micros=timestamp_proto.nanos // 1000)
```
Comment (lukecwik) [inline sdks/python/apache_beam/utils/timestamp.py]: I think we should fail here if we ever perform any rounding with a TODO stating that we should support nanos.

File: sdks/python/apache_beam/utils/timestamp.py (hunk 1 above)
Comment (rohdesamuel) [reply sdks/python/apache_beam/utils/timestamp.py]: Done.

File: sdks/python/apache_beam/utils/timestamp.py (hunk 2 above)
Comment (rohdesamuel) [reply sdks/python/apache_beam/utils/timestamp.py]: Ack, it now raises a ValueError with a reference to a JIRA as well as a TODO comment.

**Comment by rohdesamuel** [comment]
> It would also be useful to cover the duration proto as well since both are used. Do you mind doing it in this or a follow-up PR?

I don't mind, it's a small enough change.

**Comment by rohdesamuel** [comment]
Thanks for the merge!

**Pull request merged**